includes/abstracts/llms-abstract-controller-user-engagements.php

```diff
@@ -0,0 +1,252 @@
+		$is_syncing_one = $action === $actions['sync_one'];
+
+		if ( empty( $engagement_id ) ) {
+			if ( $is_syncing_one ) {
+				$code    = "llms-sync-missing-awarded-{$this->engagement_type}-id";
+				$message = $this->get_text( self::TEXT_SYNC_MISSING_AWARDED_ENGAGEMENT_ID );
+			} else {
+				$code    = "llms-sync-missing-{$this->engagement_type}-template-id";
+				$message = $this->get_text( self::TEXT_SYNC_MISSING_ENGAGEMENT_TEMPLATE_ID );
+			}
+			$result = new WP_Error( $code, $message );
+		} else {
+			if ( $is_syncing_one ) {
+				$result = $this->sync_awarded_engagement( $engagement_id );
+			} else {
+				$result = $this->sync_awarded_engagements( $engagement_id );
+			}
+		}
+
+		if ( is_wp_error( $result ) ) {
+			( new LLMS_Meta_Box_Award_Engagement_Submit() )->add_error( $result );
+		}
+
+		return $result;
+	}
+
+	/**
+	 * Sync an awarded engagement with its template.
+	 *
+	 * If the sync is successful, the {@see llms_redirect_and_exit()} function is called and this method does not return.
+	 *
+	 * @since [version]
+	 *
+	 * @param int $engagement_id Awarded engagement id.
+	 * @return void|WP_Error
+	 */
+	private function sync_awarded_engagement( $engagement_id ) {
+
+		if ( ! current_user_can( 'edit_post', $engagement_id ) ) {
+			$variables = compact( 'engagement_id' );
+			return new WP_Error(
+				"llms-sync-awarded-{$this->engagement_type}-insufficient-permissions",
+				$this->get_text( self::TEXT_SYNC_AWARDED_ENGAGEMENT_INSUFFICIENT_PERMISSIONS, $variables ),
+				$variables
+			);
+		}
+
+		$sync = $this->get_user_engagement( $engagement_id, true )->sync();
+		if ( ! $sync ) {
+			$variables = compact( 'engagement_id' );
+			return new WP_Error(
+				"llms-sync-awarded-{$this->engagement_type}-invalid-template",
+				$this->get_text( self::TEXT_SYNC_AWARDED_ENGAGEMENT_INVALID_TEMPLATE, $variables ),
+				$variables
+			);
+		}
+
+		$redirect_url = get_edit_post_link( $engagement_id, 'raw' );
+		$redirect_url = add_query_arg( 'message', 1, $redirect_url );
+		llms_redirect_and_exit( $redirect_url );
+	}
+
+	/**
+	 * Sync all the awarded engagements with their template.
+	 *
+	 * If the preflight checks are successful, the {@see llms_redirect_and_exit()} function is called after the sync is
+	 * triggered and this method does not return.
+	 *
+	 * @since [version]
+	 *
+	 * @param int $user_engagement_template_id User engagement template ID.
+	 * @return void|WP_Error
+	 */
+	private function sync_awarded_engagements( $user_engagement_template_id ) {
+
+		if ( ! current_user_can( get_post_type_object( "llms_my_{$this->engagement_type}" )->cap->edit_posts ) ) {
+			return new WP_Error(
+				"llms-sync-awarded-{$this->engagement_type}s-insufficient-permissions",
+				$this->get_text( self::TEXT_SYNC_AWARDED_ENGAGEMENTS_INSUFFICIENT_PERMISSIONS )
+			);
+		}
+
+		/**
+		 * Fires an action to trigger the bulk sync of awarded engagements.
+		 *
+		 * @since [version]
```

Comment:
```suggestion
		 * Fires an action to trigger the bulk sync of awarded engagements.
		 *
		 * The dynamic portion of this hook, `{$this->engagement_type}`, refers to the type of awarded engagement, either "achievement" or "certificate".
		 *
		 * @since [version]
```